### PR TITLE
Symfony3.4 fixes

### PR DESCRIPTION
**Pull request by niceblex33**

### CODE REVIEW

File: src/Badoo/SoftMocks.php [hunk 1]
```diff
@@ -819,6 +819,14 @@ public static function ignoreFunction($function)
         SoftMocksTraverser::ignoreFunction(ltrim($function, '\\'));
     }
 
+    /**
+     * @param string $function - Allow to mock $function
+     */
+    public static function allowFunction($function)
```
Comment (YuriyNasretdinov) [inline src/Badoo/SoftMocks.php]: You can already override internal functions if you really need to by calling SoftMocks::setRewriteInternal(true).

File: src/Badoo/SoftMocks.php [hunk 2]
```diff
@@ -2298,6 +2316,31 @@ public function rewriteScalar_MagicConst_File()
         );
     }
 
+    public function rewriteExpr_New(\PhpParser\Node\Expr\New_ $newNode)
+    {
+        if ($newNode->class == 'ReflectionClass') {
+            return new \PhpParser\Node\Expr\New_(
```
Comment (YuriyNasretdinov) [inline src/Badoo/SoftMocks.php:2411]: This approach is not desirable for several reasons:

1. You change the (pretty frequently used) class name that is created. Code that prints class names or examines the class name in any way will fail.
2. With this approach you will only be able to rewrite expressions like `new ReflectionClass();` and not `$class = 'ReflectionClass'; new $class();`. The latter might cause some subtle bugs as well.

Sadly my https://github.com/badoo/soft-mocks/pull/35 wasn't merged (yes, it is huge so I probably should've split it :)) but it should solve the same problem that you are trying to solve: it fixes reflection getFileName() as well as it allows xdebug to work and a bunch of other stuff by rewriting the file name in the file stream wrapper class.

File: src/init_with_composer.php [hunk 3]
```diff
@@ -53,6 +53,7 @@
         '/vendor/sebastian/diff/' => '/vendor/sebastian/diff/',
         '/vendor/nikic/php-parser/' => '/vendor/nikic/php-parser/',
         '/vendor/symfony/polyfill' => '/vendor/symfony/polyfill',
+        '/vendor/guzzlehttp/' => '/vendor/guzzlehttp/',
```
Comment (YuriyNasretdinov) [inline src/init_with_composer.php]: There is `SoftMocks:: addIgnoreSubPath()` method that you need to use instead. The list provided here should only ignore things that SoftMocks depends on in order to operate (that cannot be mocked) and guzzlehttp does not look like one of them.

**Review by YuriyNasretdinov — CHANGES_REQUESTED** [review]

File: src/Badoo/SoftMocks.php (hunk 2 above)
Comment (YuriyNasretdinov) [reply src/Badoo/SoftMocks.php:2411]: See discussion in https://github.com/badoo/soft-mocks/pull/6 for more details on why SoftMocks does not allow to mock `new` operator.

Comment (niceblex33) [reply src/Badoo/SoftMocks.php:2411]: Hi. Thank you for the answer. I was looking over forks and issues for sometime, but wasn't looking at merge requests at all:( I'll take a look at it.

File: src/Badoo/SoftMocks.php (hunk 1 above)
Comment (niceblex33) [reply src/Badoo/SoftMocks.php]: The idea was we need only couple function to redefine, but the library redefines many of them.
For us it's easier to say which functions to allow.

**Comment by niceblex33** [comment]
So I'll close this MR now and for now and look at your branch

File: src/Badoo/SoftMocks.php (hunk 1 above)
Comment (YuriyNasretdinov) [reply src/Badoo/SoftMocks.php]: Yeah but you can (should) define your own wrappers on top of SoftMocks already and you can apply filters on that level. I think that there should be better reasons to add new public methods than this.

**Pull request closed without merging**